overlord/snapstate/snapmgr.go

```diff
@@ -405,22 +549,40 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 }
 
 func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
-	t.State().Lock()
+	st := t.State()
+
+	// Hold the lock for the full duration of the task here so
+	// nobody observes a world where the state engine and
+	// the file system are reporting different things.
+	st.Lock()
+	defer st.Unlock()
+
 	ss, snapst, err := snapSetupAndState(t)
-	t.State().Unlock()
 	if err != nil {
 		return err
 	}
 
 	// No need to undo "snaps" in state here. The only chance of
 	// having the new state there is a working doLinkSnap call.
```

Comment:
This comment (which I wrote) seems super bogus. Let's please drop it. If something past this task fails, the new snaps will be in place. Ideally we should undo what link-snap does as best as we can.